### PR TITLE
Normalize Grant Numbers in Pass Support

### PR DESCRIPTION
**Pull request by tsande16**
Normalize the grant award numbers in pass support. The implementation is done in the pass-data-client module. The two known use cases for it is when loading grants into pass-core in the grant-loader and loading the award numbers from PMC using the nihms-loader. This will ensure that the grant award numbers are normalized in the following format: [A-Z0-9]{3}\s[A-Z0-9]{8} e.g. K23 HL153778

The normalizer will attempt to normalize grant numbers that are close to the standard format, but there are some cases which cannot be normalized for example:

- "K23 HL153778-1A1" is a known grant number that should normalize to: "K23 HL153778"
 
However

- "K23 HL15353434778" is not a known grant number and does not follow the pattern for a standard grant number

Ticket: https://github.com/eclipse-pass/main/issues/659

### CODE REVIEW

**Comment by markpatton** [comment]
It looks like DoE award numbers do not follow this structure, see https://pamspublic.science.energy.gov/WebPAMSExternal/interface/awards/AwardSearchExternal.aspx
So the normalization should be flexible for the different classes of award numbers.

Looking through how the NIH defines grant numbers, https://www.era.nih.gov/erahelp/Commons/Commons/understandGrantNums.htm, I see that year and extension fields are stripped off. But are these things that the user wants to see when shown the grant number? We may want to search ignoring those fields, but the user may want to see them. If that is the case, then we would either have to live with prefix searching and try to make the database optimize appropriately or perhaps add a new field that contains the normalized award number for searching.

**Comment by tsande16** [comment]
@markpatton Ah yes, the DOE are different than the NIH ones. Perhaps we can remove the normalization from the grant loader and keep the normalization only applied to searching for grant award numbers in the nihms loader. This was the original implementation I had done in the research ticket and I can refactor it back to this, so it wouldn't be too much to do.

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java [hunk 1]
```diff
@@ -188,22 +190,6 @@ public Grant findMostRecentGrantByAwardNumber(String awardNumber) throws IOExcep
         PassClientResult<Grant> grantResult = passClient.selectObjects(grantSelector);
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java]: Done.

File: pass-nihms-loader/nihms-pass-client/src/test/java/org/eclipse/pass/client/nihms/NihmsPassClientServiceIT.java [hunk 2]
```diff
@@ -63,28 +63,138 @@ public void lookupJournalByIssn() throws IOException {
...
     }
+
+    /**
+     * Using different variants of the same award number, demonstrate that normalized award numbers are found using
+     * different variants of an award number e.g. R01AR074846 should match R01AR074846-01A1 or 1R01AR074846-01
+     * the activity code, institute code and serial number are the minimum set of strings required to match
+     */
+    @Test
+    public void shouldFindNihGrantAwardNumber() throws IOException {
+        Grant grant1 = new Grant("1");
```
Comment (markpatton) [inline pass-nihms-loader/nihms-pass-client/src/test/java/org/eclipse/pass/client/nihms/NihmsPassClientServiceIT.java:80]: Interesting that the id is set here. It's not needed and if elide actually creates the objects with that id you might run into collisions with other tests.

Comment (tsande16) [reply pass-nihms-loader/nihms-pass-client/src/test/java/org/eclipse/pass/client/nihms/NihmsPassClientServiceIT.java:80]: Done.

**Review by markpatton — APPROVED** [review]
The changes make sense. Tests pass for me locally.

**Pull request merged**